### PR TITLE
Ecore delegates

### PR DESCRIPTION
**Pull request by arcanefoam**
Adds Ecore Delegates (validation, setting and invocation) support using EOL/EVL.

This allows for constraints, derived features and operation implementations to be defined directly in the ecore metamodel.

The documentation for use has been added to the website: https://github.com/eclipse/epsilon-website/pull/15

### CODE REVIEW

**Comment by arcanefoam** [comment]
Not sure why I am getting a conflict with the lexer, will check it out.

File: tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_AllReprotsTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.eclipse.emf.ecore.resource.impl.ResourceSetImpl;
+import org.eclipse.epsilon.ecore.delegates.test.employee.Employee;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Parameterized.class)
+public class Employee_Operation_AllReprotsTest {
```
Comment (kolovos) [inline tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_AllReprotsTest.java:22]: Class names should be in camel case for consistency e.g. EmployeeOperationAllReportsTest (also notice small typo in Repotrs).

File: tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_AllReprotsTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+	@BeforeClass
+	public static void loadModel() throws IOException {
+		InputStream stream = Employee_Operation_AllReprotsTest.class.getResourceAsStream("/ACME.xmi");
+		ResourceSet rs = new ResourceSetImpl();
+		resource = rs.createResource(URI.createURI("ACME.xmi"));
+		resource.load(stream, null);
+	}
+	
+	@Parameters( name = "Employee:Size {0}: {1}" )
```
Comment (kolovos) [inline tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_AllReprotsTest.java:32]: Indentation is a bit off in some `data()` methods.

File: tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_ReportingChainTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Parameterized.class)
+public class Employee_Operation_ReportingChainTest {
+
+	@BeforeClass
+	public static void loadModel() throws IOException {
```
Comment (kolovos) [inline tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_ReportingChainTest.java:25]: Factoring out this common method into a superclass could be useful.

File: tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_ReportingChainTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	public Employee_Operation_ReportingChainTest(
+		String employeeId,
+		int expectedSize) {
+		super();
+		this.employee = (Employee) resource.getEObject(employeeId); 
+		this.expectedSize = expectedSize;
+	}
+
+	@Test
+	public void reporting_chain_has_correct_size() {
```
Comment (kolovos) [inline tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_ReportingChainTest.java:50]: Method names should be in camel case

File: tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_ReportsToTest.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		String managerId,
+		boolean reportsTo) {
+		super();
+		this.employee = (Employee) resource.getEObject(employeeId); 
+		this.manager = (Employee) resource.getEObject(managerId); 
+		this.reportsTo = reportsTo;
+	}
+	
+	@Test
+	public void retports_to_managers() {
```
Comment (kolovos) [inline tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_ReportsToTest.java:54]: Typo: "retports"

File: tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Setting_DepartmentTest.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	public Employee_Setting_DepartmentTest(
+		String employeeId,
+		String deptId) {
+		super();
+		this.employee = (Employee) resource.getEObject(employeeId); 
+		this.dept = (Department) resource.getEObject(deptId); ;
+	}
+	
+	@Test
+	public void retports_to_managers() {
```
Comment (kolovos) [inline tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Setting_DepartmentTest.java:53]: The name of the method doesn't match its logic.

File: tests/org.eclipse.epsilon.test/META-INF/MANIFEST.MF [hunk 7]
```diff
@@ -54,4 +54,5 @@ Require-Bundle: org.eclipse.epsilon.test.dependencies,
  org.eclipse.epsilon.picto.test,
  org.eclipse.epsilon.pinset.engine.test.acceptance,
  org.eclipse.epsilon.eunit.junit.test,
- org.eclipse.epsilon.emc.cdo.test
+ org.eclipse.epsilon.emc.cdo.test,
+ org.eclipse.epsilon.ecore.delegates.test.acceptance;bundle-version="2.5.0"
```
Comment (kolovos) [inline tests/org.eclipse.epsilon.test/META-INF/MANIFEST.MF:58]: Remove bundle-version for consistency?

**Review by kolovos — COMMENTED** [review]
`org.eclipse.epsilon.ecore.delegates` should have a `pom.xml` like all other Epsilon plugins and it should be included in one of the features (org.eclipse.epsilon.emf.feature?), otherwise it won't get installed through the update site.

File: tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_AllReprotsTest.java (hunk 2 above)
Comment (arcanefoam) [reply tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_AllReprotsTest.java]: Fixed

File: tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_ReportingChainTest.java (hunk 4 above)
Comment (arcanefoam) [reply tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_ReportingChainTest.java]: Fixed

File: tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_ReportsToTest.java (hunk 5 above)
Comment (arcanefoam) [reply tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Operation_ReportsToTest.java]: Fixed

File: tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Setting_DepartmentTest.java (hunk 6 above)
Comment (arcanefoam) [reply tests/org.eclipse.epsilon.ecore.delegates.test.acceptance/src/org/eclipse/epsilon/ecore/delegates/test/acceptance/Employee_Setting_DepartmentTest.java]: Fixed

**Comment by kolovos** [comment]
Thanks Horacio. Let me know when you've had a chance to resolve the conflict in `EvlLexer` and I'll merge the PR.

**Comment by arcanefoam** [comment]
Merged

**Pull request merged**